lib/libmbedtls/core/sm2-dsa.c

```diff
@@ -0,0 +1,290 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2019-2021 Huawei Technologies Co., Ltd
+ */
+
+#include <compiler.h>
+#include <crypto/crypto.h>
+#include <mbedtls/bignum.h>
+#include <mbedtls/ecp.h>
+#include <stdlib.h>
+#include <string.h>
+#include <tee_api_types.h>
+#include <util.h>
+#include <utee_defines.h>
+
+#include "sm2-dsa.h"
+
+/* SM2 uses 256 bit unsigned integers in big endian format */
+#define SM2_INT_SIZE_BYTES 32
+
+static int rng_read(void *ignored __unused, unsigned char *buf, size_t blen)
+{
+	if (crypto_rng_read(buf, blen))
+		return MBEDTLS_ERR_MPI_FILE_IO_ERROR;
+	return 0;
+}
+
+/* Generate random number 1 <= n < max */
+static TEE_Result gen_random_upto(mbedtls_mpi *n, mbedtls_mpi *max)
+{
+	size_t sz = mbedtls_mpi_size(max);
+	bool found = false;
+	int mres = 0;
+
+	do {
+		mres = mbedtls_mpi_fill_random(n, sz, rng_read, NULL);
+		if (mres)
+			return TEE_ERROR_BAD_STATE;
+		if (mbedtls_mpi_bitlen(n) != 0 &&
+		    mbedtls_mpi_cmp_mpi(n, max) == -1)
+			found = true;
+	} while (!found);
+	return TEE_SUCCESS;
+}
+
+/*
+ * GM/T 0003.1‒2012 Part1 2 Section 6.1
+ */
+TEE_Result sm2_mbedtls_dsa_sign(uint32_t algo __unused, struct ecc_keypair *key,
+				const uint8_t *msg, size_t msg_len,
+				uint8_t *sig, size_t *sig_len)
+{
+	TEE_Result res = TEE_SUCCESS;
+	mbedtls_ecp_group grp = { };
+	mbedtls_ecp_point x1y1p = { };
+	//ecc_key ltc_key = { };
+	int mres = 0;
+	mbedtls_mpi k = { };
+	mbedtls_mpi e = { };
+	mbedtls_mpi r = { };
+	mbedtls_mpi s = { };
+	mbedtls_mpi tmp = { };
+
+	if (*sig_len < 2 * SM2_INT_SIZE_BYTES) {
+		*sig_len = 64;
+		return TEE_ERROR_SHORT_BUFFER;
+	}
+
+	mbedtls_mpi_init(&k);
+	mbedtls_mpi_init(&e);
+	mbedtls_mpi_init(&r);
+	mbedtls_mpi_init(&s);
+	mbedtls_mpi_init(&tmp);
+
+	mbedtls_ecp_point_init(&x1y1p);
+
+	mbedtls_ecp_group_init(&grp);
+	mres = mbedtls_ecp_group_load(&grp, MBEDTLS_ECP_DP_SM2);
+	if (mres) {
+		res = TEE_ERROR_GENERIC;
+		goto out;
+	}
+
+	/*
+	 * Steps A1 and A2 are the generation of the hash value e from user
+	 * information (ZA) and the message to be signed (M). There are not done
+	 * here since @msg is expected to be the hash value e already.
+	 */
+
+	/* Step A3: generate random number 1 <= k < n */
+A3:
+	res = gen_random_upto(&k, &grp.N);
+	if (res)
+		goto out;
+
+	res = TEE_ERROR_BAD_STATE;
+
+	/* Step A4: compute (x1, y1) = [k]G */
+
+	mres = mbedtls_ecp_mul(&grp, &x1y1p, &k, &grp.G, rng_read, NULL);
+	if (mres)
+		goto out;
+
+	/* Step A5: compute r = (e + x1) mod n */
+
+	mbedtls_mpi_read_binary(&e, (unsigned char *)msg, msg_len);
+	mres = mbedtls_mpi_add_mpi(&r, &e, &x1y1p.X);
+	if (mres)
+		goto out;
+	mres = mbedtls_mpi_mod_mpi(&r, &r, &grp.N);
+	if (mres)
+		goto out;
+
+	/* Step A5 (continued): return to A3 if r = 0 or r + k = n */
+
+	mres = mbedtls_mpi_add_mpi(&tmp, &r, &k);
+	if (mres)
+		goto out;
+	if (!mbedtls_mpi_cmp_int(&r, 0) || !mbedtls_mpi_cmp_mpi(&tmp, &grp.N))
+		goto A3;
```

Comment:
Please avoid making loops with goto. A `do { } while (...);` should do the job here.

Reply:
Will fix.